File: app/src/main/java/com/checkin/app/checkin/Data/RetrofitLiveData.java
package com.checkin.app.checkin.Data;

import android.util.Log;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class RetrofitLiveData<T> extends LiveData<ApiResponse<T>> implements Callback<T> {
    private final Call<T> mCall;
    private boolean mResponseDispatched = false;

    public RetrofitLiveData(Call<T> call) {
        mCall = call;
    }

    @Override
    protected void onActive() {
        super.onActive();
        if (!mResponseDispatched) {
            if (!mCall.isExecuted() && !mCall.isCanceled())
                mCall.enqueue(this);
            else
                mCall.clone().enqueue(this);
        }
    }

    @Override
    protected void onInactive() {
        super.onInactive();
        if (shouldCancel()) {
            cancel();
        }
    }

    protected void cancel() {
        if (!mCall.isCanceled())
            Log.d("Http Calls", "Call cancelled!");
        mCall.cancel();
    }

    protected boolean shouldCancel() {
        return !hasObservers();
    }

    @Override
    public void onResponse(@NonNull Call<T> call, @NonNull Response<T> response) {
        postValue(new ApiResponse<>(response));
        mResponseDispatched = true;
    }

    @Override
    public void onFailure(@NonNull Call<T> call, @NonNull Throwable t) {
        postValue(new ApiResponse<>(t));
    }
}

File: app/src/main/java/com/checkin/app/checkin/data/network/RetrofitLiveData.kt
package com.checkin.app.checkin.data.network

import android.util.Log
import androidx.lifecycle.LiveData
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

open class RetrofitLiveData<T>(private val mCall: Call<T>) : LiveData<ApiResponse<T>>() {
    private var mResponseDispatched = false
    private val mCallback: Callback<T> = object : Callback<T> {
        override fun onResponse(call: Call<T>, response: Response<T>) {
            postValue(ApiResponse(response))
            mResponseDispatched = true
        }

        override fun onFailure(call: Call<T>, t: Throwable) {
            postValue(ApiResponse(t))
        }
    }
    private val shouldCancel: Boolean
        get() = !hasObservers()

    override fun onActive() {
        super.onActive()
        if (!mResponseDispatched) {
            if (!mCall.isExecuted && !mCall.isCanceled) mCall.enqueue(mCallback) else mCall.clone().enqueue(mCallback)
        }
    }

    override fun onInactive() {
        super.onInactive()
        if (shouldCancel) cancel()
    }

    protected fun cancel() {
        if (!mCall.isCanceled) Log.d("Http Calls", "Call cancelled!")
        mCall.cancel()
    }
}